src/software/ai/navigator/path_planner/trajectory_path.h

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include "software/ai/navigator/path_planner/kinematic_constraints.h"
+#include "software/ai/navigator/path_planner/trajectory_path_node.h"
+
+/**
+ * TrajectoryPath represents a list of 2D trajectories that are connected end-to-end
+ * to form a path. A TrajectoryNode is a 2D trajectory and the time at which it ends
+ * and the next TrajectoryNode begins.
+ */
+class TrajectoryPath : public Trajectory2D
+{
+   public:
+    TrajectoryPath() = delete;
+
+    /**
+     * Constructor
+     *
+     * @param initial_trajectory The initial trajectory of this trajectory path
+     */
+    TrajectoryPath(const BangBangTrajectory2D& initial_trajectory);
```

Comment:
Why are `TrajectoryPath` and `TrajectoryPathNode` tightly coupled to `BangBangTrajectory2D`? The name implies that it will support any `Trajectory`

Reply:
Yep, you're right. Ideally it should support any trajectory type. The difficulty right now is that `TrajectoryPath::append` also instantiates a trajectory to append to the existing one, so a type is needed to be known, or a function pointer to a generator needs to be passed in. I will have to look through this and see what makes the most sense. I'm also open to ideas.

Reply:
It seems like the append is doing too much work. Maybe you could pass the already created Trajectory as an argument to `append`, and later make the class generic. I tried something locally that seems check out but there's a lot of `template` usage.

Overall, it seems like TrajectoryPath uses the interface of trajectory since all the trajectories it contains connect to each other, so it can be considered 1 long trajectory. All sub-trajectories are actual trajectories, so they use the same interface.

Reply:
The reason we generate the new trajectory inside `append` is to gaurantee that the trajectories are actually connected/continuous (i.e. the final position/velocity at `connection_time_sec` is the same as the initial position/velocity of the next trajectory). Given this constraint, I decided to generate the trajectory inside `append`

To abstract `BangBangTrajectory2d` I decided to pass in a callback function/generator which returns the interface `std::shared_ptr<Trajectory2D>`. This way, trajectory path has no idea what the trajectory type is. 
What do you guys think? Does that make sense

Reply:
I think this is fine for now, but maybe in the future it might be good to look at refactoring this. The current implementation of `BangBangTrajectory1D` and `BangBangTrajectory2D` seems a little error-prone (`generate` overwrites the trajectory, and there's this fixed array where we're responsible for keeping track of the end index!?). Seems to me that `BangBangTrajectory1D` and `BangBangTrajectory2D` should be immutable and we need some other factory class to produce them. Then we can DI that factory into `TrajectoryPath`.

Also the data representation of a `BangBangTrajectory` doesn't really have anything to do with it being "bang-bang". It's just a trajectory composed of 1D parts, and those parts could technically form any profile.

Reply:
You bring up some great points.

I agree with you that this implementation of passing in a callback is not great. A factory would be a great long term solution, specially given that we already support 1D, 2D, and angular trajectories. 

`generate` was something I added initially, with the thought in mind that it could help with performance if we can update the existing trajectory, though this was micro-optimizing too early, and has made the API not great. Removing it, or making it private and invoked by the constructor is a good idea.

The idea about the fixed array of trajectory parts is to avoid relying on `std::vector`, since during early profiling its dynamic allocation seem to hurt the performance. I considered creating a `vector` and `reserve`ing its memory, but that didn't seem too helpful. Here's some data I collected from profiling, where it showed (`incl.` column) ~10% of the total simulated test instructions where just the allocator methods (though it's worth noting that since then I've made a lot of other optimizations, and overall a lot less trajectories are collected)
![image](https://github.com/UBC-Thunderbots/Software/assets/28585597/c7359dd5-30ab-4172-9691-7126a36f3fb0)
 
Regarding the name, we can change it. This implementation was inspired by Tigers, and they call it BangBangTrajectory, so I kept the name. Not exactly sure why it's called bang-bang, since I don't think it resembles a bang bang controller.

Reply:
re: vector performance, we could consider something like [boost::container::static_vector](https://beta.boost.org/doc/libs/1_58_0/doc/html/boost/container/static_vector.html) or [eastl::fixed_vector](https://github.com/questor/eastl/blob/master/fixed_vector.h)